### PR TITLE
New logic for STAGE_INPUT valorization

### PR DESCRIPTION
**Pull request by cosimolupo**
This is a solution to issue #68, where value assigned to key STAGE_INPUT shows circular dependencies on the user-defined list of stages present in the top-level config file.

### CODE REVIEW

File: cobrawap/__main__.py [hunk 1]
```diff
@@ -323,35 +323,36 @@ def run_stage(profile=None, stage=None, extra_args=None, **kwargs):
     # lookup stage input file
     pipeline_config_path = config_path / 'configs' / 'config.yaml'
     config_dict = load_config_file(pipeline_config_path)
-    stage_idx = locate_str_in_list(config_dict['STAGES'], stage)
-    if stage_idx is None:
+    stage_idx_local = locate_str_in_list(config_dict['STAGES'], stage)
+    stage_idx_global = locate_str_in_list([v for k,v in stages.items()], stage)
```
Comment (rgutzen) [inline cobrawap/__main__.py:327]:
```suggestion
    is_first_stage = locate_str_in_list(list(stages.values()), stage) == 0
```

File: cobrawap/__main__.py [hunk 2]
```diff
@@ -323,35 +323,36 @@ def run_stage(profile=None, stage=None, extra_args=None, **kwargs):
     # lookup stage input file
     pipeline_config_path = config_path / 'configs' / 'config.yaml'
     config_dict = load_config_file(pipeline_config_path)
-    stage_idx = locate_str_in_list(config_dict['STAGES'], stage)
-    if stage_idx is None:
+    stage_idx_local = locate_str_in_list(config_dict['STAGES'], stage)
```
Comment (rgutzen) [inline cobrawap/__main__.py:326]: keep the variable name as `stage_idx`

File: cobrawap/__main__.py [hunk 3]
```diff
@@ -323,35 +323,36 @@ def run_stage(profile=None, stage=None, extra_args=None, **kwargs):
...
-    prev_stage = config_dict['STAGES'][stage_idx-1]
-    prev_stage_config_path = get_config(config_dir=config_path / prev_stage,
-                                        config_name=f'config_{profile}.yaml',
-                                        get_path_instead=True)
-    prev_config_name = Path(prev_stage_config_path).name
-    output_name = read_stage_output(stage=prev_stage, 
-                                    config_dir=config_path, 
-                                    config_name=prev_config_name)
-    stage_input = output_path / profile / prev_stage / output_name
+    if stage_idx_local>0:
```
Comment (rgutzen) [inline cobrawap/__main__.py:336]:
```suggestion
    if not is_first_stage:
```

File: cobrawap/__main__.py [hunk 4]
```diff
@@ -323,35 +323,36 @@ def run_stage(profile=None, stage=None, extra_args=None, **kwargs):
...
+        prev_stage = [v for k,v in stages.items()][stage_idx_local-1]
```
Comment (rgutzen) [inline cobrawap/__main__.py:337]: Revert to the previous logic here.
This line tries to access the global cobrawap config with the index of the pipeline config!

**Review by rgutzen — CHANGES_REQUESTED** [review]

File: cobrawap/__main__.py (hunk 3 above)
Comment (cosimolupo) [reply cobrawap/__main__.py:336]: Checking `stage_idx_local>0` already does the job. The boolean variable `is_first_stage` is redundant.

File: cobrawap/__main__.py (hunk 4 above)
Comment (cosimolupo) [reply cobrawap/__main__.py]: This was a leftover bug from previous changes in this PR. Thanks for having pointed it out. Fixed through the last commit.

File: cobrawap/__main__.py (hunk 2 above)
Comment (cosimolupo) [reply cobrawap/__main__.py:326]: It's important to differentiate among the stage index in the custom-defined list of stages, and the larger one present in the `~/.cobrawap/config` file.

Comment (rgutzen) [reply cobrawap/__main__.py:326]: There is only one stage_idx in use here (the "local" one), so there is no need to differentiate

Comment (cosimolupo) [reply cobrawap/__main__.py:326]: Yes, but using "talking" names for variables helps... We can use "stage_idx" as it used to be, or "user_stage_idx", or any other meaningful suggestion...

**Comment by mdenker** [comment]
@rgutzen Please recheck the latest changes.

**Review by rgutzen — CHANGES_REQUESTED** [review]

File: cobrawap/pipeline/utils/Snakefile [hunk 5]
```diff
@@ -13,9 +13,11 @@ CONFIG_PATH = Path(get_setting('config_path'))
 OUTPUT_PATH = Path(get_setting('output_path'))
 
 if 'STAGE_INPUT' not in config:
-    logger.warning('No STAGE_INPUT defined for running the stage individually! '
-                   'You can set it via the command line with '
-                   '`--config STAGE_INPUT=/path/to/file`.')
+    if config['STAGE_NAME']!=get_setting('stages')['1']:
```
[the remark this answers is not in the document]
Comment (cosimolupo) [reply cobrawap/pipeline/utils/Snakefile:16]: Ok for the definition of boolean variable `is_first_stage `, but it's important to keep the two conditions separated. `config['STAGE_INPUT'] = None` has to be set whenever `STAGE_INPUT` is not defined in the config file, while the warning has to be shown only if the stage to be executed is not stage_01. I also added an extra check, so to ignore `STAGE_INPUT` value if the user has mistakenly set it while asking for the execution of stage_01.

File: cobrawap/pipeline/utils/Snakefile [hunk 6]
```diff
@@ -12,11 +12,13 @@ from utils.snakefile import get_setting
...
-    logger.warning('No STAGE_INPUT defined for running the stage individually! '
-                   'You can set it via the command line with '
-                   '`--config STAGE_INPUT=/path/to/file`.')
+is_first_stage = config['STAGE_NAME'] == get_setting('stages')['1']
+if 'STAGE_INPUT' not in config or ('STAGE_INPUT' in config and is_first_stage):
     config['STAGE_INPUT'] = None
+    if not is_first_stage:
+        logger.warning('No STAGE_INPUT defined for running stage \'{}\' individually! '.format(config['STAGE_NAME']) +
+                       'You can set it via the command line with ' +
+                       '`--config STAGE_INPUT=/path/to/file`.')
```
Comment (rgutzen) [inline cobrawap/pipeline/utils/Snakefile:21]: This logic got a bit messy, how about we simplify it like this:

```suggestion
if is_first_stage:
    config['STAGE_INPUT'] = None
if 'STAGE_INPUT' not in config:
    logger.warning('No STAGE_INPUT defined for running stage \'{}\' individually! '.format(config['STAGE_NAME']) +
                       'You can set it via the command line with ' +
                       '`--config STAGE_INPUT=/path/to/file`.')
```

**Review by rgutzen — CHANGES_REQUESTED** [review]

File: cobrawap/pipeline/utils/Snakefile (hunk 6 above)
Comment (gulpgiulia) [reply cobrawap/pipeline/utils/Snakefile]: I suggest to accept the change.
I also suggest to consider to send an alert to the user any time (as in this case) when the user's settings have been overwritten, something like: logger.warning(Cobrawap is running from first stage - ‘STAGE_INPUT forced to Nan. Your user settings have been overwritten)

**Review by rgutzen — APPROVED** [review]

**Comment by gulpgiulia** [comment]
I also suggest to consider to send an alert to the user any time (as in this case) when the user's settings have been overwritten, something like: logger.warning(Cobrawap is running from first stage - ‘STAGE_INPUT forced to Nan. Your user settings have been overwritten)

**Pull request merged**